packages/client/lib/client/RESP2/decoder.ts

```diff
@@ -0,0 +1,229 @@
+import StringComposer from './composers/string';
+
+enum Types {
+    SIMPLE_STRING = 43, // +
+    ERROR = 45, // -
+    INTEGER = 58, // :
+    BULK_STRING = 36, // $
+    ARRAY = 42 // *
+}
+
+const CR = 13; // \r
+
+export type Reply = string | Buffer | ErrorReply | number | null | Array<Reply>;
+
+type ArrayReply = Array<Reply> | null;
+
+export type ReturnStringsAsBuffers = () => boolean;
+
+interface RESP2Options {
+    returnStringsAsBuffers: ReturnStringsAsBuffers;
+    onReply(reply: Reply): unknown;
+}
+
+interface ArrayInProcess {
+    array: Array<Reply>;
+    pushCounter: number;
+}
+
+export default class RESP2Decoder {
+    #options: RESP2Options;
+
+    constructor(options: RESP2Options) {
+        this.#options = options;
+    }
+
+    #cursor = 0;
+
+    #type?: Types;
+
+    #bufferComposer = new BufferComposer();
+
+    #stringComposer = new StringComposer();
+
+    #composer: BufferComposer | StringComposer = this.#stringComposer;
+
+    write(chunk: Buffer): void {
+        while (this.#cursor < chunk.length) {
+            if (!this.#type) {
+                this.#composer = this.#options.returnStringsAsBuffers() ?
+                    this.#bufferComposer :
+                    this.#stringComposer;
+
+                this.#type = chunk[this.#cursor];
+                if (++this.#cursor >= chunk.length) break;
+            }
+
+            const reply = this.#parseType(chunk, this.#type);
+            if (reply === undefined) break;
+
+            this.#type = undefined;
+            this.#options.onReply(reply);
+        }
+
+        this.#cursor -= chunk.length;
+    }
+
+    #parseType(chunk: Buffer, type: Types, arraysToKeep?: number): Reply | undefined {
+        switch (type) {
+            case Types.SIMPLE_STRING:
+                return this.#parseSimpleString(chunk, this.#composer);
+
+            case Types.ERROR:
+                return this.#parseError(chunk);
+
+            case Types.INTEGER:
+                return this.#parseInteger(chunk);
+
+            case Types.BULK_STRING:
+                return this.#parseBulkString(chunk);
+
+            case Types.ARRAY:
+                return this.#parseArray(chunk, arraysToKeep);
+        }
+    }
+
+    #parseSimpleString<
+        C extends Composer<T>,
+        T = C extends Composer<infer TT> ? TT : never
+    >(
+        chunk: Buffer,
+        composer: C
+    ): T | undefined {
+        const crIndex = this.#findCRLF(chunk);
+        if (crIndex !== -1) {
+            const reply = composer.end(
+                chunk.slice(this.#cursor, crIndex)
+            );
+            this.#cursor = crIndex + 2;
+            return reply;
+        }
+
+        const toWrite = chunk.slice(this.#cursor);
+        composer.write(toWrite);
+        this.#cursor = chunk.length;
+    }
+
+    #findCRLF(chunk: Buffer): number {
+        for (let i = this.#cursor; i < chunk.length; i++) {
+            if (chunk[i] === CR) {
+                return i;
```

Comment:
Might be a style difference. In python I'm more likely to do something like:
```
f = -1
for thing ... <my loop>:
    if things match:
        f = thing
        break
return f
```

I find it makes it easier to read functions when you're not looking for multiple returns